From 981b30377a45e32fb6529771e7cb6acbf2f0e033 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 01:01:55 +0100
Subject: [PATCH 75/86] urlmon: Avoid calling remove_dot_segments with too
 small buffer.

=================================================================
==888==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffffe1fee2a at pc 0x6ffff83ca415 bp 0x7ffffe1fd4d0 sp 0x7ffffe1fd518
READ of size 2 at 0x7ffffe1fee2a thread T0
04a8:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff83ca414 in remove_dot_segments /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/uri.c:606:37
    #1 0x6ffff83bfb11 in parse_canonicalize /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/uri.c:6593:30
    #2 0x6ffff83befe2 in CoInternetParseIUri /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/uri.c:6963:14
    #3 0x0001400a59a4 in test_CoInternetParseIUri_InvalidArgs /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/uri.c:11478:14
    #4 0x0001400a59a4 in func_uri /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/uri.c:12219:5
    #5 0x00014011cde0 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #6 0x00014011cde0 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #7 0x00014011ec8f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #8 0x6ffffc1c4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #9 0x6ffffc39fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1fee2a is located in stack of thread T0 at offset 5738 in frame
    #0 0x00014008974f in func_uri /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/uri.c:12124

  This frame has 100 object(s):
    [32, 40) 'session.i2942' (line 11247)
    [64, 72) 'uri.i2710' (line 12034)
    [96, 104) 'ubf.i' (line 12035)
    [128, 136) 'ps.i' (line 12036)
    [160, 168) 'ub.i' (line 12037)
    [192, 200) 'bstr.i' (line 12038)
    [224, 228) 'dword.i' (line 12039)
    [240, 244) 'eq.i' (line 12040)
    [256, 264) 'ui.i' (line 12041)
    [288, 296) 'mon.i' (line 11711)
    [320, 328) 'base_mon.i' (line 11711)
    [352, 360) 'uri.i2586' (line 11713)
    [384, 392) 'base_uri.i' (line 11713)
    [416, 424) 'uri.i2499' (line 11557)
    [448, 848) 'result.i2500' (line 11562)
    [912, 916) 'result_len.i2501' (line 11563)
    [928, 936) 'base.i2446' (line 11394)
    [960, 968) 'result.i2447' (line 11399)
    [992, 1000) 'received.i2448' (line 11410)
    [1024, 1032) 'base.i2391' (line 11263)
    [1056, 1064) 'relative.i2392' (line 11268)
    [1088, 1096) 'result.i2393' (line 11273)
    [1120, 1128) 'received.i2394' (line 11284)
    [1152, 1160) 'session.i' (line 11231)
    [1184, 1192) 'uri.i2278' (line 11520)
    [1216, 5386) 'result.i2279' (line 11528)
    [5648, 5652) 'result_len.i2280' (line 11529)
    [5664, 5672) 'uri.i2091' (line 11426)
    [5696, 5702) 'tmp.i2092' (line 11427)
    [5728, 5732) 'result.i2093' (line 11429)
    [5744, 5752) 'base.i1889' (line 11302) <== Memory access at offset 5738 underflows this variable
    [5776, 5784) 'result.i1890' (line 11302)
    [5808, 5816) 'received.i1891' (line 11353)
    [5840, 5844) 'received172.i' (line 11371)
    [5856, 5864) 'base.i' (line 11012)
    [5888, 5896) 'relative.i' (line 11012)
    [5920, 5928) 'result.i1721' (line 11012)
    [5952, 5960) 'received.i1722' (line 11065)
    [5984, 5988) 'received180.i' (line 11082)
    [6000, 6008) 'uri.i1581' (line 10925)
    [6032, 6040) 'factory.i' (line 10926)
    [6064, 6072) 'builder.i1582' (line 10927)
    [6096, 6104) 'tmp.i' (line 10959)
    [6128, 6136) 'result.i1583' (line 10960)
    [6160, 6164) 'result_len.i' (line 10961)
    [6176, 6184) 'tmp114.i' (line 10995)
    [6208, 6216) 'uri.i1534' (line 10897)
    [6240, 6248) 'builder.i1535' (line 10901)
    [6272, 6276) 'has.i' (line 10906)
    [6288, 6292) 'port.i1536' (line 10907)
    [6304, 6312) 'builder.i1404' (line 10812)
    [6336, 6344) 'uri.i1405' (line 10847)
    [6368, 6376) 'result.i' (line 10868)
    [6400, 6408) 'received.i1406' (line 10876)
    [6432, 6440) 'builder.i1283' (line 10675)
    [6464, 6472) 'uri.i1284' (line 10681)
    [6496, 6504) 'test.i1285' (line 10689)
    [6528, 6536) 'builder.i1168' (line 10570)
    [6560, 6568) 'uri.i1169' (line 10576)
    [6592, 6596) 'received.i1170' (line 10577)
    [6608, 6616) 'prop.i' (line 10596)
    [6640, 6644) 'len.i1171' (line 10597)
    [6656, 6664) 'builder.i1087' (line 9634)
    [6688, 6696) 'received.i1088' (line 9640)
    [6720, 6724) 'len.i' (line 9641)
    [6736, 6740) 'port.i' (line 9641)
    [6752, 6756) 'set.i' (line 9642)
    [6768, 6776) 'builder.i982' (line 9542)
    [6800, 6808) 'test.i983' (line 9548)
    [6832, 6840) 'uri.i984' (line 9548)
    [6864, 6872) 'builder.i' (line 9356)
    [6896, 6904) 'uri.i938' (line 9357)
    [6928, 6936) 'uri.i867' (line 9321)
    [6960, 6968) 'received.i868' (line 9335)
    [6992, 7000) 'uri.i824' (line 9307)
    [7024, 7032) 'uri.i794' (line 9284)
    [7056, 7076) 'fragmentW.i' (line 9285)
    [7120, 7136) 'custom_uri.i' (line 9206)
    [7152, 7160) 'uriA.i' (line 9207)
    [7184, 7192) 'uriB.i' (line 9207)
    [7216, 7220) 'equal.i' (line 9208)
    [7232, 7240) 'uri.i540' (line 8934)
    [7264, 7268) 'received.i542' (line 8966)
    [7280, 7288) 'uri.i476' (line 8883)
    [7312, 7316) 'received.i478' (line 8908)
    [7328, 7336) 'uri.i389' (line 8772)
    [7360, 7364) 'received.i390' (line 8780)
    [7376, 7380) 'receivedLen.i' (line 8806)
    [7392, 7400) 'uri.i317' (line 8683)
    [7424, 7428) 'received.i318' (line 8718)
    [7440, 7448) 'uri.i241' (line 8363)
    [7472, 7476) 'received.i242' (line 8370)
    [7488, 7492) 'received49.i' (line 8397)
    [7504, 7512) 'uri.i195' (line 8287)
    [7536, 7544) 'received.i' (line 8295)
    [7568, 7576) 'received88.i' (line 8337)
    [7600, 7608) 'uri.i165' (line 8271)
    [7632, 7640) 'uri.i139' (line 8243)
    [7664, 7680) 'invalidW.i' (line 8245)
    [7696, 7704) 'uri.i' (line 8232)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/uri.c:606:37 in remove_dot_segments
Shadow bytes around the buggy address:
  0x7ffffe1feb80: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fec00: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fec80: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f2
  0x7ffffe1fed00: f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ffffe1fed80: f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f8 f2 00 f2 f2 f2
=>0x7ffffe1fee00: 06 f2 f2 f2 04[f2]f8 f2 f2 f2 f8 f2 f2 f2 f8 f2
  0x7ffffe1fee80: f2 f2 f8 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2
  0x7ffffe1fef00: f8 f2 f2 f2 f8 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2
  0x7ffffe1fef80: f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f8 f2 f2 f2
  0x7ffffe1ff000: f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f8 f2 f8 f2 f2 f2
  0x7ffffe1ff080: f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==888==ABORTING
04a8:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:381588: dlls/urlmon/tests/x86_64-windows/uri.ok] Fehler 1
---
 dlls/urlmon/uri.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/urlmon/uri.c b/dlls/urlmon/uri.c
index cb5266518d1..d8e6b60dffc 100644
--- a/dlls/urlmon/uri.c
+++ b/dlls/urlmon/uri.c
@@ -6588,7 +6588,7 @@ static HRESULT parse_canonicalize(const Uri *uri, DWORD flags, LPWSTR output,
     /* Sometimes the path is the very last component of the IUri, so
      * see if the dot segments need to be reduced now.
      */
-    if(reduce_path && path) {
+    if(reduce_path && path && len < output_len) {
         DWORD current_path_len = (output+len) - path;
         DWORD new_path_len = remove_dot_segments(path, current_path_len);
 
-- 
2.47.1

